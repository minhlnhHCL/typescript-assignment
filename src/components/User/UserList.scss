.btn-create {
    cursor: pointer;
    padding: 8px;
    width: 100%;
    border-radius: 8px;
    border: 0;
    background: transparent;
    color: black;
    box-shadow: rgba(0, 0, 0, 0.15) 1.95px 1.95px 2.6px;
}

.link-div {
    text-decoration: none;
    color: black;
    font-size: 1.25rem;
}


.list-container {
    margin: 0 auto;
    max-width: 480px;
    padding: 0 16px;

    .list-wrapper {
        display: flex;
        flex-direction: column;
    }
}