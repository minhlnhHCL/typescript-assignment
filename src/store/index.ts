import { configureStore, PreloadedState } from '@reduxjs/toolkit'
import { useDispatch, useSelector } from 'react-redux'
import { combineReducers } from 'redux'

import users from './userSlice'
import modal from './modalSlice'
import toast from './toastSlice'

const reducer = combineReducers({
    users,
    modal,
    toast
})
const setupStore = (preloadedState?: PreloadedState<RootState>) => {
    return configureStore({
        reducer,
        preloadedState,
        middleware: (getDefaultMiddleware) =>
            getDefaultMiddleware({
                serializableCheck: false
            })
    })
}

export type RootState = ReturnType<typeof reducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']

export default setupStore
