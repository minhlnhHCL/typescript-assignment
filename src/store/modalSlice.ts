import { createSlice, PayloadAction } from "@reduxjs/toolkit"


export interface IModal {
    open: boolean,
    title: string,
    children: JSX.Element | string | null
    successCb: () => void

}

const initialState: IModal = {
    open: false,
    title: '',
    children: null,
    successCb: () => { },
}

export const modalSlice = createSlice({
    name: 'modal',
    initialState,
    reducers: {
        setModal: (state, action: PayloadAction<IModal>) => {
            state.children = action.payload.children
            state.open = action.payload.open
            state.title = action.payload.title
            state.successCb = action.payload.successCb
        }
    }
})

export const { setModal } = modalSlice.actions
export default modalSlice.reducer