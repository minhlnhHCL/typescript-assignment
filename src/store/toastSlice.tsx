import { createSlice, PayloadAction } from "@reduxjs/toolkit"

export interface IToastDetails {
    status: string,
    msg: string
}

export interface IToast {
    notification: IToastDetails | null
}

const initialState: IToast = {
    notification: null
}

const toastSLice = createSlice({
    name: 'toast',
    initialState,
    reducers: {
        showToast: (state, action: PayloadAction<IToastDetails>) => {
            state.notification = action.payload
        },
        hideToast: (state) => {
            state.notification = null
        }
    }
})

export const { showToast, hideToast } = toastSLice.actions

export default toastSLice.reducer