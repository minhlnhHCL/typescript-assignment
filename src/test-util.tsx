import React, { PropsWithChildren, ReactNode } from 'react'
import { render } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import type { PreloadedState } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'

import setupStore from 'store'
import type { AppStore, RootState } from 'store/index'
import { LoadingContext, LoadingProvider } from 'context/LoadingContext'
import { BrowserRouter } from 'react-router-dom'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
  onLoading: () => void,
  offLoading: () => void,
}

export * from '@testing-library/react'

export const renderWithRedux = (
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    onLoading = () => { },
    offLoading = () => { },
    ...renderOptions
  }: ExtendedRenderOptions = { onLoading: () => { }, offLoading: () => { } }
) => {
  const Wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => {
    return <BrowserRouter>
      <Provider store={store}>
        <LoadingContext.Provider value={{ offLoading, onLoading }}>
          {children}
        </LoadingContext.Provider>
      </Provider></BrowserRouter>
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}