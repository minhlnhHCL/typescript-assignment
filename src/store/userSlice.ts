import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { IUser } from '../components/User/UserInterface'
export const GET_USER = 'GET_USER'
export const ADD_USER = 'ADD_USER'
export const EDIT_USER = 'EDIT_USER'
export const DELETE_USER = 'DELETE_USER'
export const CURRENT_USER = 'CURRENT_USER'

export interface UserState {
    users: IUser[],
}

export const initialState: UserState = {
    users: [],
}
//UserSlice

export const userSLice = createSlice({
    name: 'users',
    initialState,
    reducers: {
        getUsers: (state, action: PayloadAction<IUser[]>) => {
            if (state.users.length == 1) {
                const newUserArr = [...state.users, ...action.payload]
                state.users = newUserArr
                return state
            }
            state.users = action.payload
        },
        addUser: (state, action: PayloadAction<IUser>) => {
            const index = state.users.findIndex(el => el.id == action.payload.id);
            if (index == -1) {
                state.users = [action.payload, ...state.users]
                return state
            } else {
                const newPayload = { ...action.payload, id: action.payload.id && action.payload.id + 1 }
                state.users = [newPayload, ...state.users]
                return state
            }
        },
        editUser: (state, action: PayloadAction<IUser>) => {
            const index = state.users.findIndex(el => el.id == action.payload.id);
            const newUserArr = [...state.users]
            newUserArr[index] = action.payload
            state.users = newUserArr
        },
        deleteUser: (state, action: PayloadAction<IUser>) => {
            const clonedArr = [...state.users]
            const newUserArr = clonedArr.filter(item => item.id !== action.payload.id)
            state.users = newUserArr
        }
    }
})


export const { getUsers, editUser, addUser, deleteUser } = userSLice.actions
export default userSLice.reducer