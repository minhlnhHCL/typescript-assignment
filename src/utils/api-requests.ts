import axios from 'axios';
import { IUser } from '../components/User/UserInterface';

export interface IApiRequest {
  url?: string | '',
  body?: IUser
}

export const baseUrl = 'https://dummyjson.com/users';

export const get = ({url}: IApiRequest) => {
  return axios.get(baseUrl + url);
};

export const post = ({ url, body }: IApiRequest) => {
  return axios.post(baseUrl + url, body);
};

export const put = ({ url, body }: IApiRequest) => {
  return axios.put(baseUrl + url, body);
};

export const deleteReq = ({ url }: IApiRequest) => {
  return axios.delete(baseUrl + url);
};

