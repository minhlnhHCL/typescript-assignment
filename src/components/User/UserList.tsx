import { useEffect } from 'react'
import './UserList.scss'
import { useNavigate } from 'react-router-dom'
import { getUsers } from 'store/userSlice'
import { get } from 'utils/api-requests'
import { useAppDispatch, useAppSelector } from 'store/hooks'
import UserWrapper from './UserWrapper/UserWrapper'
import { IUser } from './UserInterface'
import { useLoadingContext } from 'context/LoadingContext'

const UserList = () => {
    const { users } = useAppSelector(state => state.users)
    const dispatch = useAppDispatch()
    const navigate = useNavigate()
    const {onLoading, offLoading} = useLoadingContext()

    useEffect(() => {
        onLoading()
        const getData = async () => {
            try {
                if (users.length === 0 || users.length === 1) {
                    const res = await get({url: ''})
                    if (res.status === 200) {
                        return dispatch(getUsers(res.data.users))
                    }
                }
            } catch (error) {
                console.log(error)
            } finally {
                offLoading()
            }
        }
        getData()
    }, [users])

    return (
        <div className='list-container mt-32'>
            <button className='btn-create' onClick={() => {
                navigate('/create')
            }}>Create New User</button>
            <div className='list-wrapper'>
                {
                    users !== undefined && users.length !== 0 && users.map((item: IUser, idx) => (
                        <UserWrapper key={`${item.id}-${idx}`} data={item} />
                    ))
                }
            </div>
        </div>
    )
}

export default UserList