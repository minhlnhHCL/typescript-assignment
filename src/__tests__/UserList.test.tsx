import { renderWithRedux } from "test-util"
import UserList from "components/User/UserList"
import { useNavigate } from "react-router-dom"
import { cleanup, waitFor } from '@testing-library/react'
import { baseUrl, get } from "utils/api-requests"
import { useAppDispatch, useAppSelector } from "store/hooks"
import type { AppDispatch } from 'store/index'
import { useDispatch } from "react-redux"

const sampleUsers = [{
    id: 1,
    firstName: "Minh",
    lastName: "Luong",
    image: 'https://robohash.org/hicveldicta.png',
    age: 25,
    company: {
        title: "Front End Engineer",
    },
}, {
    id: 2,
    firstName: "Min",
    lastName: "Min",
    image: 'https://robohash.org/hicveldicta.png',
    age: 21,
    company: {
        title: "Front End Engineer",
    },
}]


jest.mock('utils/api-requests')

jest.mock("store/hooks", () => ({
    ...jest.requireActual("react-redux"),
    useAppSelector: jest.fn(),
    useAppDispatch: jest.fn(),
}))

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn(),
}))

describe('User List', () => {
    beforeEach(() => {
        (useAppDispatch as jest.Mock).mockImplementationOnce(() => jest.fn());
        (useNavigate as jest.Mock).mockReturnValue(() => { });
        (useAppSelector as jest.Mock).mockImplementation(() => { });
        (get as jest.Mock).mockResolvedValue({
            data: {
                users: []
            }
        });
    })

    afterAll(cleanup)

    it('should run its side effect', async () => {
        const onLoading = jest.fn();
        const offLoading = jest.fn();
            (useAppSelector as jest.Mock).mockReturnValue({
                users: []
            });
        (get as jest.Mock).mockResolvedValueOnce({
            data: {
                users: sampleUsers
            }
        });
        const { container } = renderWithRedux(<UserList />, { offLoading, onLoading });
        expect(get).toBeCalled();
        await waitFor(() => expect(onLoading).toBeCalled());
        expect(useAppDispatch).toBeCalled();

        await waitFor(() => expect(offLoading).toBeCalled());
        expect(container).toMatchSnapshot();

    })
})