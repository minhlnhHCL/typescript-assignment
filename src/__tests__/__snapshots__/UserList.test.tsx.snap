// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User List should run its side effect 1`] = `
<div>
  <div
    class="list-container mt-32"
  >
    <button
      class="btn-create"
    >
      Create New User
    </button>
    <div
      class="list-wrapper"
    />
  </div>
</div>
`;
